### PR TITLE
Expand useTracking API to accept contextual data

### PR DESCRIPTION
**Pull request by bgergen**
This PR could probably use more refinement, but I want to get some early feedback on it. It addresses issue https://github.com/nytimes/react-tracking/issues/138. I spent some time considering this API vs having the `Track` component take in props. I think this is the better option because it allows contextual data to be passed into the hook and get back a `trackEvent` function that can make use of that data, which would not be possible if the data had to be passed into the `Track` component instead.

One minor (I think) concern I still have is that it might be too easy for a developer to forget to wrap their markup in the `Track` component, and mistakenly believe that child components are receiving contextual data when they're not. Open question: Would it make any sense to emulate the `useState` API, and do something like:
```js
const [Track, tracking] = useTracking(data);
```
That would at least force the developer to very clearly opt out of using `Track` by needing to do:
```js
const [, tracking] = useTracking(data);
```
Maybe that's too heavy handed though. I don't know—something worth thinking about perhaps.

### CODE REVIEW

File: src/__tests__/hooks.test.js
Comment (bgergen) [inline src/__tests__/hooks.test.js:491]: @tizmagik Not sure whether this was what you were intending, but this was what I needed to do to match the expected test output.

File: src/__tests__/hooks.test.js
Comment (bgergen) [inline src/__tests__/hooks.test.js:833]: Another option here (and maybe the more idiomatic option?) would be for `Page` to have a button or something that calls `executeAction` on click, and then using enzyme to simulate the click event, rather than calling it manually ourselves.

File: src/__tests__/hooks.test.js
Comment (bgergen) [inline src/__tests__/hooks.test.js:885]: This is kind of a weird one. I _think_ that calling the ref's `focus` method programmatically does not result in the element's `onFocus` method being called. In any case, I'm not so sure that this test is really testing anything useful here—in the version for the decorator, it's testing that `track` forwards the ref along like we'd expect, but here the hook doesn't actually interact with the ref in any way.

File: src/__tests__/hooks.test.js
Comment (bgergen) [reply src/__tests__/hooks.test.js:57]: Yup, will do!

File: src/__tests__/e2e.test.js
Comment (bgergen) [reply src/__tests__/e2e.test.js:714]: This changed because I moved the context export into the `useTracking` file so that there wouldn't be a circular reference with the decorator importing and calling the hook and the hook importing the context object.

File: src/__tests__/hooks.test.js
Comment (bgergen) [reply src/__tests__/hooks.test.js:885]: Yup, I think that's right. I'll remove this test.

File: src/withTrackingComponentDecorator.js
Comment (bgergen) [reply src/withTrackingComponentDecorator.js:57]: Yeah, this adds an additional component wrapping the decorated component. I think this should be avoidable by having both the HoC and the useTracking hook both call a separate shared hook that will return the context value. Then the useTracking hook can return its `Track` component, and the HoC can wrap its decorated component in the provider. I'll make this update—it should mostly just involve breaking the meat of the useTracking functionality out into its own hook.

**Comment by bgergen** [comment]
@tizmagik Yes, I think you're right about that tuple API. It could be a bit awkward, and I let my head drift off and forget that such a change wouldn't be backward compatible. Also, I wouldn't want to make things more complicated in anticipation of a potential issue that we haven't actually seen users have in the wild. If we're hearing that folks are running into that issue then we can reevaluate. In the meantime, I'll make that update to prevent any changes in the HoC API and do some manual testing to make sure everything looks good!

File: src/__tests__/hooks.test.js
Comment (bgergen) [inline src/__tests__/hooks.test.js:837]: For some reason I needed to add this line to wait for the state to update for this test but not for the following one. My guess is that it has to do with the way the functions within the component are written—it's kind of baffling to me, but it seems to have to do with React's internals rather than our code and I don't think it has any actual implications for this PR—just a JS/React quirk.

File: src/__tests__/useTracking.test.js
Comment (bgergen) [inline src/__tests__/useTracking.test.js:9]: Removed this test because calling `useTracking` without the tree being wrapped in `track` no longer throws an error.

File: src/__tests__/hooks.test.js
Comment (tizmagik) [reply src/__tests__/hooks.test.js:837]: Maybe we have to use `act()` or something? https://reactjs.org/docs/test-utils.html#act

File: src/__tests__/useTracking.test.js
Comment (tizmagik) [reply src/__tests__/useTracking.test.js:9]: Makes sense -- but maybe we should restore this test and just assert that it now _doesn't_ throw?

File: src/__tests__/hooks.test.js
Comment (tizmagik) [inline src/__tests__/hooks.test.js:57]: I _think_ this should still be expected to pass even if you don't return `<Track />` yea?

```suggestion
      return null;
```

**Review by tizmagik — APPROVED** [review]
This is brilliant, thanks for doing this!

Two minor things I commented on inline, but I leave it up to you -- I'm fine merging as-is.

One other thing we'll need to do is add some new docs around this. Happy to leave that for a follow-on PR or just create an issue but we'll def need a new section relatively high up on the current README explaining the new Hooks-only capability.

File: src/__tests__/hooks.test.js
Comment (bgergen) [reply src/__tests__/hooks.test.js:837]: Hmmm mayyyybe. I'll give it a shot, though usually jest yells at you to use it if it's necessary. The thing is, I've had to do this before with other tests to wait for the component to re-render with the new state, so that doesn't surprise me all that much—what does surprise me though is that the following test _doesn't_ seem to need to wait for a re-render. Anyway, I don't think it's something worth spending too much time on, but I will try using `act` to see if that helps.

**Comment by bgergen** [comment]
@tizmagik No prob, it's been fun! Going to make the suggested changes to the tests, and then I think I can take a stab at updating the docs in a separate PR if that seems alright.

File: src/__tests__/hooks.test.js
Comment (bgergen) [reply src/__tests__/hooks.test.js:837]: Ahhhh, we _did_ need `act`! But the error wasn't printing because we had previously mocked `console.error` and didn't unmock. Updating.

Comment (tizmagik) [reply src/__tests__/hooks.test.js]: Ah, good catch!

**Comment by bgergen** [comment]
FYI, I found a bug while doing some manual testing. Many of the functions in the hook have a dependency on `trackingData`, which means that if the consumer passes in an object literal and the component renders more than once, the hook will read it as a new object being passed in and all of the memoization in the hook is worthless. The same is true for the `dispatchOnMount` function if that's defined inline. I need to figure out the best way to memoize those 2 arguments—feels too onerous to ask the consumer of the hook to do so before passing them in—and write or update a couple of tests to catch this. Probably won't be able to get to that until later in the week though.

**Comment by tizmagik** [comment]
Ah, good catch. I think [this test](https://github.com/nytimes/react-tracking/pull/168/files#diff-59b49daedd74542afad8ae14f6823ea97a3fdc12ced5ab9f00cbc74d998553d2R612) attempts to get at that (unnecessary renders), but I think wouldn't catch the case you're talking about.

File: src/__tests__/hooks.test.js
Comment (tizmagik) [inline src/__tests__/hooks.test.js:670]: What you have here is functionally the same but just in case this behavior ever changes do you mind updating this so that the actual `dispatch` calls are counted and not just the side effect of calling dispatchOnMount that's counted?

```suggestion
      useTracking(
        {},
        {
          dispatch,
          dispatchOnMount: () => {
            return { test: true };
          },
        }
      );
```

Then later:

```js
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenLastCalledWith({
      test: true,
    });
```

File: src/__tests__/hooks.test.js
Comment (tizmagik) [inline src/__tests__/hooks.test.js:689]: Maybe click it a couple times and wrap in `act()` for good measure? Would be great to also assert that the state is actually updated to confirm.

**Review by tizmagik — CHANGES_REQUESTED** [review]
Great! One minor thing then I think we can ship this! 💪

File: src/__tests__/hooks.test.js
Comment (bgergen) [reply src/__tests__/hooks.test.js]: Great call on this. Adding a couple extra clicks exposed a bug. I decided not to wrap the clicks in `act()` because it wasn't making a difference in the result, and it turns out that's because as of React 16.8, if we use `mount`, enzyme [wraps calls](https://enzymejs.github.io/enzyme/#reacttestutilsact-wrap) to `.simulate` in `act` for us. I think we just needed it in the other spot to handle the asynchronicity.

I still want to do some manual testing again before we merge this just to be sure everything is working properly. Should we aim for a Monday merge?

**Review by tizmagik — APPROVED** [review]

**Pull request merged**
